src/common/sv2_noise.h

```diff
@@ -0,0 +1,295 @@
+ *  https://noiseprotocol.org/noise.html
+ *  Revision 38, 2018-07-11
+ *
+ *  Stratum v2 handshake and cipher specification:
+ *  https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md
+ */
+
+/** Section 3: All Noise messages are less than or equal to 65535 bytes in length. */
+static constexpr size_t NOISE_MAX_CHUNK_SIZE = 65535;
+/** Sv2 spec 4.5.2 */
+
+/** Simple certificate for the static key signed by the authority key.
+ * See 4.5.2 and 4.5.3 of the Stratum v2 spec.
+ */
+class Sv2SignatureNoiseMessage
+{
+public:
+    /** Size of a Schnorr signature. */
+    static constexpr size_t SCHNORR_SIGNATURE_SIZE = 64;
+    /** Size of serialized message, which does not include the static key.  */
+    static constexpr size_t SIZE = 2 + 4 + 4 + SCHNORR_SIGNATURE_SIZE;
+
+private:
+    uint16_t m_version = 0;
+    uint32_t m_valid_from = 0;
+    uint32_t m_valid_to = 0;
+    std::array<unsigned char, SCHNORR_SIGNATURE_SIZE> m_sig;
+
+    /** Hash of version, valid from/to and the static key. */
+    uint256 GetHash();
+    void SignSchnorr(const CKey& authority_key, Span<unsigned char> sig);
+
+public:
+    Sv2SignatureNoiseMessage() = default;
+    Sv2SignatureNoiseMessage(uint16_t version, uint32_t valid_from, uint32_t valid_to, const XOnlyPubKey& static_key, const CKey& authority_key);
+
+    /* The certificate serializes pubkeys in x-only format, not EllSwift. */
+    XOnlyPubKey m_static_key = {};
+
+    [[nodiscard]] bool Validate(XOnlyPubKey authority_key);
+
+    template <typename Stream>
+    // The static_key is signed for, but not serialized.
+    void Serialize(Stream& s) const
+    {
+        s << m_version
+          << m_valid_from
+          << m_valid_to
+          << m_sig;
+    }
+    template <typename Stream>
+    void Unserialize(Stream& s)
+    {
+        s >> m_version >> m_valid_from >> m_valid_to >> m_sig;
+    }
+};
+
+/*
+ * The CipherState uses m_key (k) and m_nonce (n) to encrypt and decrypt ciphertexts.
+ * During the handshake phase each party has a single CipherState, but during
+ * the transport phase each party has two CipherState objects: one for sending,
+ * and one for receiving.
+ *
+ * See chapter "5. Processing rules" of the Noise paper.
+ */
+class Sv2CipherState
+{
+public:
+    static constexpr size_t HASHLEN{32};
+
+    Sv2CipherState() = default;
+    explicit Sv2CipherState(uint8_t key[HASHLEN]);
+
+    /** Decrypt message
+     * @param[in] associated_data associated data
+     * @param[in] ciphertext message with encrypted and authenticated chunks.
+     * @param[out] plain message (defragmented)
+     * @returns whether decryption succeeded
+     */
+    [[nodiscard]] bool DecryptWithAd(Span<const std::byte> associated_data, Span<std::byte> ciphertext, Span<std::byte> plain);
+    /** Encrypt message
+     * @param[in] associated_data associated data
+     * @param[out] plain message
+     * @param[in] ciphertext message with encrypted and authenticated chunks.
```

Comment:
nit: Should the `ciphertext` message be the @param[out] in the EncryptWithAdd method?